User facing documentation for Logstash on ECK (#6585)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/logstash.asciidoc

Reviewer: This file needs to be linked in  `orchestrating-elastic-stack-applications.asciidoc`

Reviewer 2: Done with 6977d5dd81f755732512601b66509d8156087398

:page_id: logstash
:logstash_recipes: https://raw.githubusercontent.com/elastic/cloud-on-k8s/{eck_release_branch}/config/recipes/logstash
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Run Logstash on ECK

experimental[]

This section describes how to configure and deploy Logstash with ECK.

* <<{p}-logstash-quickstart,Quickstart>>
* <<{p}-logstash-configuration,Configuration>>
** <<{p}-logstash-configuring-logstash,Configuring Logstash>>
** <<{p}-logstash-pipelines,Configuring Pipelines>>
** <<{p}-logstash-pipelines-es,Using Elasticsearch in Logstash Pipelines>>
** <<{p}-logstash-expose-services,Exposing Services>>
* <<{p}-logstash-configuration-examples,Configuration examples>>
* <<{p}-logstash-advanced-configuration,Advanced Configuration>>
** <<{p}-logstash-jvm-options,Setting JVM Options>>
* <<{p}-logstash-technical-preview-limitations,Technical Preview Limitations>>

> NOTE: Running Logstash on ECK is compatible only with Logstash 8.7+.

Reviewer 3: Do we need to remove this? I do see LS 8.6.1 versions are in settings (ex: [logstash.yml](https://github.com/elastic/cloud-on-k8s/blob/feature/logstash/config/samples/logstash/logstash.yaml#L7)) and as I understood stack monitoring for Logstash requires 8.7.0+. 
Or, we can change all configs version to 8.7.0 to align with this note.

[id="{p}-logstash-quickstart"]
== Quickstart

> Add the following specification to create a minimal logstash deployment that will listen to a beats agent or elastic agent configured to send to logstash on port 5044, create the service and write the output to an Elasticsearch cluster named `quickstart`, created in the link:k8s-quickstart.html[Elasticsearch quickstart].

Reviewer 4: I see one more `logstash` without capital.
```suggestion
Add the following specification to create a minimal Logstash deployment that will listen to a Beats agent or Elastic Agent configured to send to Logstash on port 5044, create the service and write the output to an Elasticsearch cluster named `quickstart`, created in the link:k8s-quickstart.html[Elasticsearch quickstart].
```

NAME              AVAILABLE   EXPECTED   AGE   VERSION
quickstart        3           3          4s    {version}
----

. List all the Pods that belong to a given Logstash specification.
+
[source,sh]
----
kubectl get pods --selector='logstash.k8s.elastic.co/name=quickstart'
----
+
[source,sh]
----
NAME              READY   STATUS    RESTARTS   AGE
quickstart-ls-0   1/1     Running   0          91s
----

. Access logs for a pod.
+
[source,sh]
----
kubectl logs -f quickstart-ls-0
----

[id="{p}-logstash-configuration"]
== Configuration

[id="{p}-logstash-upgrade-specification"]
=== Upgrade the Logstash specification

You can upgrade the Logstash version or change settings by editing the YAML specification. ECK applies the changes by performing a rolling restart of Logstash Pods.

[id="{p}-logstash-configuring-logstash"]
=== Logstash configuration

The Logstash configuration (equivalent to logstash.yml) is defined in the `spec.config` section:

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
  - name: quickstart
    clusterName: qs

>   config:
>     pipeline.workers: 4  <1>
>     log.level: debug
> ----
> <1> Customize logstash configuration using logstash.yml settings here

Reviewer 4: ```suggestion
  config: <1>
    pipeline.workers: 4  
    log.level: debug
----
<1> Customize Logstash configuration using `logstash.yml` settings here
```

        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }
----

Alternatively, it can be provided through a Secret specified in the `spec.pipelinesRef` element. The Secret must have a `logstash.yml` entry with this configuration:
[source,yaml,subs="attributes,+macros"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
    - clusterName: qs
      name: quickstart
  pipelinesRef:
    secretName: quickstart-pipeline
---
apiVersion: v1
kind: Secret
metadata:
  name: quickstart-pipeline
stringData:
  pipelines.yml: |-
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }

----

> Logstash on ECK will* support all options present in `pipelines.yml`, including settings to update the number of workers, and

Author: Removed

 the size of the batch that the pipeline will process. This also includes using `path.config` to point to volumes

>  mounted on the logstash container:

Reviewer 4: ```suggestion
 mounted on the Logstash container:
```

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
    - clusterName: qs
      name: quickstart
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }
----

> NOTE: Persistent Queues and Dead Letter Queues will be supported in a later release, but are not currently supported.

Reviewer 2: ```suggestion
NOTE: Persistent Queues and Dead Letter Queues are not currently supported.
```
Best practice: do not take uncertain commitments.

[id="{p}-logstash-pipelines-es"]
=== Using Elasticsearch in Logstash Pipelines

The `spec.elasticsearchRefs` section provides a mechanism to help configure Logstash to estabish a secured connection to one or more managed Elasticsearch clusters. By default, it targets all nodes in your cluster. If you want to direct traffic to specific nodes of your Elasticsearch cluster, refer to <<{p}-traffic-splitting>> for more information and examples.

> In order to use `elasticsearchRefs` in a logstash pipeline, the logstash operator will create the necessary resources from the associated elasticsearch, and provide environment variables to allow these resources to be accessed from pipeline configuration, and will be replaced at runtime with the appropriate values.

Reviewer 2: ```suggestion
To use `elasticsearchRefs` in a Logstash pipeline, the Logstash operator will create the necessary resources from the associated Elasticsearch, and provides environment variables to allow these resources to be accessed from the pipeline configuration, and will be replaced at runtime with the appropriate values.
```
You might want to split this into smaller sentences. It's too long and a bit confusing.

The environment variables have a fixed naming convention:
`NORMALIZED_CLUSTERNAME_ES_HOSTS`
`NORMALIZED_CLUSTERNAME_ES_USER`
`NORMALIZED_CLUSTERNAME_ES_PASSWORD`
`NORMALIZED_CLUSTERNAME_ES_SSL_CERTIFICATE_AUTHORITY}`

where NORMALIZED_CLUSTERNAME is the value taken from the `clusterName` field of the `elasticsearchRef` property, capitalized, and `-` transformed to `_` - eg, prod-es, would becomed PROD_ES.

> NOTE: The `clusterName` value should be unique across namespaces.

Author: Good point, will reword for clarity

> NOTE: The logstash ECK operator will create a user called `eck_logstash_user_role` when an `elasticsearchRef` is specified. This user has the following permissions:

Reviewer 2: ```suggestion
NOTE: The Logstash ECK operator will create a user called `eck_logstash_user_role` when an `elasticsearchRef` is specified. This user has the following permissions:
```

The following example demonstrates how to create a Logstash deployment that connects to
different elasticsearch instances, one of which is in a separate namespace:

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:        <1>
    - clusterName: prod-es  <2>
      name: prod
    - clusterName: qa-es    <3>
      name: qa
      namespace: qa
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {   <4>
            hosts => [ "${PROD_ES_ES_HOSTS}" ]
            user => "${PROD_ES_ES_USER}"
            password => "${PROD_ES_ES_PASSWORD}"
            cacert => "${PROD_ES_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
          elasticsearch {   <4>
            hosts => [ "${QA_ES_ES_HOSTS}" ]
            user => "${QA_ES_ES_USER}"
            password => "${QA_ES_ES_PASSWORD}"
            cacert => "${QA_ES_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }

----

> <1> Define Elasticsearch references in the CRD. This will create the appropriate secrets to store certificate details and the rest of the connection information, and create environment variables to allow them to be referred to in Logstash pipeline configurations.
> <2> This refers to an Elasticsearch cluster residing in the same namespace as the logstash instances
> <3> This refers to an Elasticsearch cluster residing in a different namespace to the logstash instance
> <4> Elasticsearch output definitions - use the environment variables created by the Logstash operator when specifying an `ElasticsearchRef`. Note the use of "normalized" versions of the `clusterName` in the environment variables used to populate the relevant fields

Reviewer 4: ```suggestion
<1> Define Elasticsearch references in the CRD. This will create the appropriate Secrets to store certificate details and the rest of the connection information, and create environment variables to allow them to be referred to in Logstash pipeline configurations.
<2> This refers to an Elasticsearch cluster residing in the same namespace as the Logstash instances.
<3> This refers to an Elasticsearch cluster residing in a different namespace to the Logstash instances.
<4> Elasticsearch output definitions - use the environment variables created by the Logstash operator when specifying an `ElasticsearchRef`. Note the use of "normalized" versions of the `clusterName` in the environment variables used to populate the relevant fields.
```

[id="{p}-logstash-expose-services"]
=== Expose Services

> By default, the Logstash operator creates a headless service for the metrics endpoint to enable metric collection by the metricbeat sidecar for stack monitoring:

Reviewer 4: ```suggestion
By default, the Logstash operator creates a headless Service for the metrics endpoint to enable metric collection by the Metricbeat sidecar for Stack Monitoring:
```

> +

Author: Removed

* Particular care should be taken with plugins that:
** Retrieve data from external sources.
*** Plugins that retrieve data from external sources, and require some level of coordination between nodes to split up work, are not good candidates for scaling horizontally, and would likely produce some data duplication. These are plugins such as the JDBC input plugin, which has no automatic way to split queries across logstash instances, or the S3 input, which has no way to split which buckets to read across logstash instances.
*** Plugins that retrieve data from external sources, where work is distributed externally to logstash, but may impose their own limits. These are plugins like the kafka input, or azure event hubs, where the parallelism is limited by the number of partitions vs the number of consumers. In cases like this, extra logstash pods may be idle if the number of consumer threads multiplied by the number of pods is greater than the number of partitions.
** Plugins that require events to be received in order.

> *** Certain plugins, such as the aggregate filter, expect events to be received in strict order to run without error or data loss. Any plugin that requires the number of pipeline workers to be `1` will also have issues when horizontal scaling is used.

Author: I've removed the first level of bullets

* If the pipeline does not contain any such plugin, the number of Logstash instances can be increased by setting the `count` property in the Logstash resource:

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 3
----

[id="{p}-logstash-technical-preview-limitations"]
== Technical Preview Limitations

Note that this release is a technical preview, is still under active development and has additional limitations:

* No integrated support for persistence
** The operator provides no integrated support for persistence, including PQ, DLQ support and plugins that may require persistent storage to keep track of state - any persistence should be added manually.
* `ElasticsearchRef` implementation in plugins is in preview mode
** Adding elasticsearch to plugin definitions requires the use of environment variables populated by the Logstash operator, which may change in future versions of the logstash operator.

> * Limited support for plugins. Note that this is not an exhaustive list, and plugins outside of the logstash plugin matrix have not been considered for this list:

Reviewer 4: ```suggestion
* Limited support for plugins. Note that this is not an exhaustive list, and plugins outside of the Logstash plugin matrix have not been considered for this list:
```

          limits:
            memory: 4Gi
            cpu: 2
----

The name of the container in the Pod template must be `logstash`.

[id="{p}-logstash-configuration-examples"]
== Configuration examples

This section contains manifests that illustrate common use cases, and can be your starting point in exploring Logstash deployed with ECK. These manifests are self-contained and work out-of-the-box on any non-secured Kubernetes cluster. They all contain a three-node Elasticsearch cluster and a single Kibana instance.

CAUTION: The examples in this section are for illustration purposes only and should not be considered to be production-ready. Some of these examples use the `node.store.allow_mmap: false` setting on Elasticsearch which has performance implications and should be tuned for production workloads, as described in <<{p}-virtual-memory>>.

=== Single Pipeline defined in CRD

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-eck.yaml
----

Deploys Logstash with a single pipeline defined in the CRD

=== Single Pipeline defined in CRD

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-eck.yaml
----

> Deploys Logstash with a single pipeline defined in the CRD

Author: It is not 🤦
